gordo_components/model/metadata.py

```diff
@@ -0,0 +1,137 @@
+        Variable with associated thread.
+    tls._metadata: dict
+        Attribute of local variable to store metadata to.
+    metadata: dict
+        Helper attribute referencing `tls._metadata` for use in context manager.
+    """
+
+    def __init__(
+        self, multiples: Optional[bool] = False, overwrite: Optional[bool] = False
+    ):
+        if multiples and overwrite:
+            raise ValueError(
+                "Both `overwrite` and `multiples` cannot be set `True`"
+            )
+
+        self._tls = threading.local()
+        self._tls._metadata = {}
+        self._multiples = multiples
+        self._overwrite = overwrite
+
+        self.metadata = self._tls._metadata
+
+    def log_meta(self, key: str, value: Any):
+        """
+        Log a metadata key/value pair
+
+        Parameters
+        ----------
+        key: str
+            The key to add to the metadata dictionary.
+        value: Any
+            The value to add to the metadata dictionary.
+        """
+        if key in self.metadata:
+            if self._overwrite:
+                warnings.warn(f"Value for key {key} has been overwritten.")
+                self.metadata.update({key: value})
+            if self._multiples:
+                self.metadata.update(
+                    {MetadataCollector.increment(self.metadata, key): value}
+                )
+        else:
+            self.metadata.update({key: value})
+
+    @staticmethod
+    def increment(d: dict, key: str) -> str:
+        """
+        Add sequential suffix if key exists
+
+        Parameters
+        ----------
+        d: dict
+            Dictionary to check for existing key.
+        key: str
+            Key to check if exists in dictionary.
+
+        Returns
+        -------
+        key: str
+            Key with integer suffix added.
+
+        Examples
+        --------
+        >>> d = {}
+        >>> key = increment(d, "test_key")
+        >>> print(key)
+        "test_key"
+        >>> key = increment(d, "test_key")
+        >>> print(key)
+        "test_key_1"
+        """
+        if key in d:
+            try:
+                i = int(key.split("_")[-1])
+            except ValueError:
+                i = 1
+            except Exception as e:
+                raise e
+            key += f"_{i}"
+            return MetadataCollector.increment(d, key)
+        else:
+            return key
+
+
+@contextmanager
+def autometa(*args, **kwargs):
+    """
+    Metadata collecting context manager
+
+    See Also
+    --------
+    MetaCollector.log_meta: Threaded logger of key/value pairs to metadata
+
+    Examples
+    --------
+    >>> with autometa() as mc:
+            mc.log_meta("test_key", 0)
+            print(mc.metadata["test_data"])
+    0
+    """
+    mc = MetadataCollector(*args, **kwargs)
+    try:
+        yield mc
+    except Exception as e:
+        raise e
+    finally:
+        # Remove collecter object for Thread cleanliness
+        del mc
```

Comment:
How much does this actually do? It would be dropped as it goes out of scope/reference count locally, but any client code which holds a reference to it would become the owner, the same way this code is valid:
```python
>>>  a = 'hi'
>>>  b = a
>>> del a
>>> b
'hi'
```
perhaps I misunderstand.

Reply:
Implementation is now different, and there is a test now to check that the thread local data is removed.